### PR TITLE
📊 1006 Clinical Data GQL

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

- Adds Resolvers for Clinical Data + Clinical Errors
  - nested ClinicalData / ClinicalErrors resolver structure is removed
- Updates to Clinical Errors:
  - Adds async handlers for Dictionary Validation
  - Makes ClinicalErrors non-nullable
  - Removes nested resolver + repeated async methods
- Adds Convert to GQL methods

Related changes:
Gateway: https://github.com/icgc-argo/platform-api/pull/672
UI: https://github.com/icgc-argo/platform-ui/pull/2713

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [x] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [inline src/clinical/clinical-service.ts:413]: Revalidating records requires multiple dictionary validations

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [inline src/clinical/clinical-service.ts:47]: Made Type names more specific to their purpose + return value

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [inline src/clinical/clinical-service.ts:63]: Added new types for GQL request variables

File: src/dictionary/manager.ts
Comment (demariadaniel) [inline src/dictionary/manager.ts:303]: Async only works properly in new `For of` loops

Comment (ciaranschutte) [reply src/dictionary/manager.ts:303]: unclear where the async function is. Was this never working?

File: src/dictionary/manager.ts
Comment (ciaranschutte) [inline src/dictionary/manager.ts:312]: `isValid` is always going to be `true` here, so the original line is correct

File: src/clinical/clinical-service.ts
Comment (ciaranschutte) [reply src/clinical/clinical-service.ts:63]: I think `interface extends type` is incorrect usage.
The way types extend is usually by unions.
I'm not sure why TS isn't throwing an error here.
Usually the main reason to use an interface is to extend from an interface.

What's the reasoning using the `interface` vs a `type`?
I ask because most of our code is written in types based on some talk we had a while back on slack with preferring types.

File: src/clinical/clinical-service.ts
Comment (ciaranschutte) [inline src/clinical/clinical-service.ts:345]: `Donor` has a `donorId` property already
```

export interface Donor {
  _id?: string;
  __v?: number; // mongodb property not being filtered out
  createBy?: string;
  schemaMetadata: SchemaMetadata;
  donorId?: number;
  gender: string;
  submitterId: string;
  programId: string;
  specimens: Array<Specimen>;
  clinicalInfo?: ClinicalInfo;
  primaryDiagnoses?: Array<PrimaryDiagnosis>;
  familyHistory?: Array<FamilyHistory>;
  comorbidity?: Array<Comorbidity>;
  followUps?: Array<FollowUp>;
  treatments?: Array<Treatment>;
  exposure?: Array<Exposure>;
  biomarker?: Array<Biomarker>;
  createdAt?: string;
  updatedAt?: string;
  completionStats?: CompletionStats;
}
``` from src/clinical/clinical-entities.ts

File: src/clinical/clinical-service.ts
Comment (ciaranschutte) [inline src/clinical/clinical-service.ts:390]: Will `donorId` ever not be a number? 
If it's always a number no need to have an additional iteration with `filter`

File: src/clinical/clinical-service.ts
Comment (ciaranschutte) [inline src/clinical/clinical-service.ts:413]: a comment would be helpful here. there is a lot to follow with multiple Promise arrays and nested mapping and then just returning the 1st index place of the array.

File: src/schemas/clinical-resolvers/clinicalData.ts
Comment (ciaranschutte) [inline src/schemas/clinical-resolvers/clinicalData.ts:29]: what is the double `[][]` my google skills are failing me?

File: src/schemas/clinical-resolvers/clinicalData.ts
Comment (ciaranschutte) [inline src/schemas/clinical-resolvers/clinicalData.ts:34]: having `name` as either a `string | false` seems like a type nightmare waiting to happen?

**Review by ciaranschutte — APPROVED** [review]
@demariadaniel left some questions & suggestions

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [reply src/clinical/clinical-service.ts:345]: But it's `donorId?` which is what's causing the issue

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [reply src/clinical/clinical-service.ts:390]: Apparently 🤷 Donor is typed with a `?`, so from Typescript's perspective no, donorId will not always be a number
I will dig into this a bit

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [reply src/clinical/clinical-service.ts:413]: Good call, will add

File: src/schemas/clinical-resolvers/clinicalData.ts
Comment (demariadaniel) [reply src/schemas/clinical-resolvers/clinicalData.ts:29]: An array containing arrays containing EntityDisplayRecords
`[ [ {} ], [ {} ], [ {} ] ]`

File: src/schemas/clinical-resolvers/clinicalData.ts
Comment (demariadaniel) [reply src/schemas/clinical-resolvers/clinicalData.ts:34]: Fair, will fix

File: src/dictionary/manager.ts
Comment (demariadaniel) [reply src/dictionary/manager.ts:303]: Let me work back through this; dictionaryService.processRecords is the root of the issue and the fact that we're sometimes revalidating hundreds of records.

Comment (demariadaniel) [reply src/dictionary/manager.ts:303]: But, may have got my wires crossed

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [reply src/clinical/clinical-service.ts:412]: Added

File: src/schemas/clinical-resolvers/clinicalData.ts
Comment (demariadaniel) [reply src/schemas/clinical-resolvers/clinicalData.ts:34]: OK updated, used a conditional `continue` to handle submitter_id
This original logic was written a long time ago, definitely worth an update

File: src/dictionary/manager.ts
Comment (demariadaniel) [reply src/dictionary/manager.ts]: Right, cleaned up

File: src/clinical/clinical-service.ts
Comment (demariadaniel) [reply src/clinical/clinical-service.ts:63]: Aha I see, still working on my Typescript. So where I got confused is `Type extends` is not a thing, `Type & {}` is the alternative. I used `interface` b/c I wanted to `extend` and didn't revisit.

File: src/clinical/clinical-service.ts
Comment (ciaranschutte) [reply src/clinical/clinical-service.ts]: I see. the filter for checking number type threw me.
either it's an optional type and you check its not null in code. or it's a mandatory value.


we can do those checks rather than change the type. so you are right with a filter, it can just be a filter checking for existence of value. or setup default values, as I've seen in other code. 

just observations, totally your call on which way you want it to work

File: src/dictionary/manager.ts
Comment (demariadaniel) [reply src/dictionary/manager.ts:303]: OK, I was able to de-Promisify across several files:
https://github.com/icgc-argo/argo-clinical/pull/1051/commits/5ee92de3cf23fd50c26925e17c9db86183550279

Debugging is a windy road. GQL was returning null values, but getClinicalErrors was returning correct values, so it appeared race condition was a possibility.

Comment (demariadaniel) [reply src/dictionary/manager.ts:303]: Much improved this way

**Review by ciaranschutte — DISMISSED** [review]

**Review by ciaranschutte — APPROVED** [review]
approving after merge. re ran tests fixed the failing build. something about a merge into head failing now? @demariadaniel  re-request if you need another approval otherwise I won't be notified

**Pull request merged**
